AvionicsSoftware-AtollicProject/Src/xtract.c

```diff
@@ -0,0 +1,103 @@
+static UART_HandleTypeDef* uart;
+//-------------------------------------------------------------------------------------------------------------------------------------------------------------
+// ENUMS AND ENUM TYPEDEFS
+//-------------------------------------------------------------------------------------------------------------------------------------------------------------
+
+//-------------------------------------------------------------------------------------------------------------------------------------------------------------
+// STRUCTS AND STRUCT TYPEDEFS
+//-------------------------------------------------------------------------------------------------------------------------------------------------------------
+
+//-------------------------------------------------------------------------------------------------------------------------------------------------------------
+// TYPEDEFS
+//-------------------------------------------------------------------------------------------------------------------------------------------------------------
+
+//-------------------------------------------------------------------------------------------------------------------------------------------------------------
+// FUNCTION PROTOTYPES
+//-------------------------------------------------------------------------------------------------------------------------------------------------------------
+//-------------------------------------------------------------------------------------------------------------------------------------------------------------
+// Description:
+//  Enter description for static function here.
+//
+// Returns:
+//  Enter description of return values (if any).
+//-------------------------------------------------------------------------------------------------------------------------------------------------------------
+//static int private_fnc(
+//    const void * param1, // Enter description of param1.
+//    const void * param2  // Enter description of param2.
+//    );
+
+//-------------------------------------------------------------------------------------------------------------------------------------------------------------
+// FUNCTIONS
+//-------------------------------------------------------------------------------------------------------------------------------------------------------------
+void vTask_xtract(void *pvParameters){
+	uart = (UART_HandleTypeDef*) pvParameters;
+
+	intro(); //display help on start up
+	char *cmd_buf = (char*) malloc(sizeof(char) * BUFFER_SIZE); //command buffer
+
+	/* As per most FreeRTOSD tasks, this task is implemented in an infinite loop. */
+	while(1){
+		transmit(uart, ">> ");
+		cmd_buf = receive_command(uart); //puts input into buffrx
+		handle_command(cmd_buf); //handles command sitting in buffrx
+	}
+} //vTaskUART_CLI END
+
+void handle_command(char* command){
+	char* output = (char*) malloc(BUFFER_SIZE * sizeof(char));
```

Comment:
In general for embedded systems you want to avoid using malloc. It can lead to memory leaks and heap fragmentation which quickly become issues with the limited memory of our microcontroller(96kb).

In this case there is a memory leak, since you never free the memory allocated to 'output'.
When I ran this task it crashed after handling approx. 40 commands, from an illegal memory access fault(basically ran out of memory).

It would probably be best to allocate the memory statically, but if you have a good reason for using malloc then checkout the freertos page on memory management. It explains about the freertos malloc implementation, and the different heap management schemes that we can use.